### PR TITLE
Make patchdd work on macOS

### PR DESCRIPTION
**Pull request by jb-gcx**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -2,19 +2,14 @@ cmake_minimum_required(VERSION 3.1.0 FATAL_ERROR)
 project(csn)
 
 option(BUILD_PATCHDD "build patchdd utility" OFF)
+option(BUILD_CSN "build main casync tool" ON)
+option(BUILD_TESTING "build casync tool tests" ON)
```
Comment (florolf) [inline CMakeLists.txt]: `include(CTest)` will create this option. Is there any reason why you had to pull this out here (and below)?

File: CMakeLists.txt [hunk 2]
```diff
@@ -25,26 +20,49 @@ add_compile_options(
...
+CHECK_INCLUDE_FILE("linux/fs.h" LINUX_FS_FOUND)
+CHECK_INCLUDE_FILE("sys/disk.h" SYS_DISK_FOUND)
+
+if(${LINUX_FS_FOUND})
+	add_compile_definitions(USE_LINUX_FS)
+elseif(${SYS_DISK_FOUND})
+	add_compile_definitions(USE_SYS_DISK)
+endif()
+
+if(${BUILD_CSN})
```
Comment (florolf) [inline CMakeLists.txt]: `${}` is not necessary here and to be honest I've not seen this in other CMake files. Is this a new style recommendation I've missed?

File: utils.c [hunk 3]
```diff
@@ -5,10 +5,46 @@
 #include <unistd.h>
 #include <sys/stat.h>
 #include <sys/ioctl.h>
-#include <linux/fs.h>
 
 #include "utils.h"
 
+
+#ifdef USE_LINUX_FS
```
Comment (florolf) [inline utils.c]: Detecting this via `CHECK_INCLUDE_FILE` feels a little fragile to me. I'd rather distinguish between the different OSes explicitly.

See for example https://github.com/samangh/gptfdisk/blob/master/diskio-unix.cc#L445-L479

File: utils.c [hunk 4]
```diff
@@ -5,10 +5,46 @@
...
+#include <linux/fs.h>
+
+static int blockdevice_size(int fd, off_t* size_ptr) {
```
Comment (florolf) [inline utils.c]: Style nitpicks:
* `off_t *size_ptr`
*  Newline before top-level blocks

**Review by florolf — CHANGES_REQUESTED** [review]
There's two changes in here that can go into separate commits:
 * Allow disabling the main casync tool
 * Make `fd_size` work on macOS

File: CMakeLists.txt (hunk 1 above)
Comment (jb-gcx) [reply CMakeLists.txt]: I didn't know this, I thought this was implicitly created somewhere :+1:

File: utils.c (hunk 3 above)
Comment (jb-gcx) [reply utils.c]: It felt like the best way to me, but I guess it's a question of taste. I changed it to just check for `__APPLE__`, which also removes quite a bit of cmake sourcery 👍

**Comment by florolf** [comment]
Merged with some minor preprocessor changes. Thanks!

**Pull request closed without merging**